File: evals/run-retrieval.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs'
import type { Chunk } from './lib/corpus.ts'
import { chunkConfigFromEnv, expandChunks } from './lib/chunk.ts'
import { CORPUS, corpusTextHash, type EmbeddingsIndex } from './lib/corpus.ts'
import { readJsonLines } from './lib/jsonl.ts'
import { tokenizeStemmed } from './lib/stem.ts'
import { ndcgAtK, pairedBootstrap, recallAtK, reciprocalRank } from './lib/metrics.ts'
import { unpackVectors } from './lib/vectors.ts'
import { bm25Retriever } from './retrievers/bm25.ts'
import { byParent } from './retrievers/by-parent.ts'
import { denseRetriever } from './retrievers/dense.ts'
import { fasttextRetriever } from './retrievers/fasttext.ts'
import { keywordRetriever, type Retriever } from './retrievers/keyword.ts'

/**
 * A label pointing at a chunk that no longer exists is not a retrieval failure — it is a stale
 * label, and scoring it as a miss quietly lowers every number with no visible cause.
 */
export function partitionLabels(labels: Labelled[], chunks: Chunk[]): { usable: Labelled[]; stale: number } {
  const ids = new Set(chunks.map((c) => c.id))
  const usable = labels.filter((l) => l.relevant.every((id) => ids.has(id)))
  return { usable, stale: labels.length - usable.length }
}

export type Labelled = {
  query: string
  relevant: string[]
  kind: 'bulk' | 'hard'
  edited?: boolean
  /** How a hard-set target was located: by typing words into the corpus, or by date/project. */
  via?: 'search' | 'browse'
}

/**
 * Two ways of writing a label bypass the leakage guard, and both would raise the numbers rather
 * than the realism if they leaked: an edited query is written with its chunk on screen, and a
 * search-located hard target is by construction a chunk containing the words that were typed.
 * Neither can be prevented outright, so each is recorded and scored apart from its complement —
 * if the exposed half scores higher, that gap is the leak, printed instead of buried in the mean.
 */
function reportSplit(
  name: string,
  subset: Labelled[],
  rest: Labelled[],
  retrievers: Retriever[],
  warning: string,
): void {
  if (subset.length === 0 || rest.length === 0) {
    return
  }
  console.log(`\n${name}: ${subset.length} of ${subset.length + rest.length}`)
  for (const r of retrievers) {
    const a = score(r, subset).row
    const b = score(r, rest).row
    console.log(
      `${r.name.padEnd(10)} ndcg@10 ${name} ${fmt(a.ndcg10, subset.length)} (n=${subset.length}) ` +
        `vs rest ${fmt(b.ndcg10, rest.length)} (n=${rest.length})`,
    )
  }
  console.log(warning)
}

type Row = {
  retriever: string
  kind: string
  n: number
  r1: number
  r3: number
  r3max: number
  r10: number
  mrr: number
  ndcg10: number
  p50ms: number
  p95ms: number
}

type Comparison = {
  better: string
  worse: string
  kind: string
  metric: string
  n: number
  meanDiff: number
  lower: number
  upper: number
}

const REMEDY = 're-run pnpm eval:embed'

/**
 * Chunk ids are content-addressed, so an id surviving does not prove the embedding matches —
 * only the text hash does. Likewise a query appended to the labels after the last
 * embed run has no vector. Both fail as a uniform zero score folded into the mean with no
 * warning — the most interesting-looking result in the table, and fabricated. Crash instead.
 */
export function assertDenseInputsFresh(
  chunks: Array<{ id: string; text: string }>,
  meta: EmbeddingsIndex,
  queries: string[],
  qv: Record<string, number[]>,
): void {
  const corpusIds = chunks.map((c) => c.id)
  const corpus = new Set(corpusIds)
  const embedded = new Set(meta.ids)
  const missing = corpusIds.filter((id) => !embedded.has(id))
  const stale = meta.ids.filter((id) => !corpus.has(id))
  if (missing.length || stale.length) {
    throw new Error(
      `embeddings.index.json does not match the corpus: ${missing.length} chunk(s) unembedded, ` +
        `${stale.length} embedded id(s) no longer in the corpus — ${REMEDY}`,
    )
  }
  // Ids can match while every chunk body has been rewritten by an ingest-rule or cap change;
  // without this the dense scores would be computed against text that no longer exists.
  if (meta.textHash !== corpusTextHash(chunks)) {
    throw new Error(
      `embeddings.index.json was built from different chunk text (same ids, ` +
        `${meta.textHash ? 'hash mismatch' : 'no hash recorded'}) — ${REMEDY}`,
    )
  }
  const bad = queries.filter((q) => (qv[q]?.length ?? 0) !== meta.dim)
  if (bad.length) {
    throw new Error(
      `${bad.length}/${queries.length} labelled queries have no vector of dim ${meta.dim} ` +
        `(first: ${JSON.stringify(bad[0])}) — ${REMEDY}`,
    )
  }
}

export function score(r: Retriever, items: Labelled[]): { row: Omit<Row, 'retriever' | 'kind'>; perQuery: number[] } {
  const mean = (xs: number[]) => (xs.length ? xs.reduce((s, x) => s + x, 0) / xs.length : 0)
  // Per-query timing: for the dense retriever this is the brute-force cosine scan, which is
  // the number that decides when a real vector index becomes necessary.
  const times: number[] = []
  const ranked = items.map((it) => {
    const t0 = performance.now()
    const out = r.search(it.query, 10)
    times.push(performance.now() - t0)
    return out
  })
  times.sort((x, y) => x - y)
  const pct = (p: number) => times[Math.min(times.length - 1, Math.floor(p * times.length))] ?? 0
  const ndcg = ranked.map((rk, i) => ndcgAtK(rk, items[i].relevant, 10))
  return {
    row: {
      n: items.length,
      r1: mean(ranked.map((rk, i) => recallAtK(rk, items[i].relevant, 1))),
      r3: mean(ranked.map((rk, i) => recallAtK(rk, items[i].relevant, 3))),
      // Recall divides by the number of relevant chunks, so a query with eight of them caps
      // recall@3 at 3/8 — pooling lowers the ceiling of the metric it was meant to fix. Printed
      // next to the score because a reader comparing 0.22 against 1.0 draws the wrong conclusion.
      r3max: mean(items.map((it) => Math.min(3, it.relevant.length) / it.relevant.length)),
      r10: mean(ranked.map((rk, i) => recallAtK(rk, items[i].relevant, 10))),
      mrr: mean(ranked.map((rk, i) => reciprocalRank(rk, items[i].relevant))),
      ndcg10: mean(ndcg),
      p50ms: pct(0.5),
      p95ms: pct(0.95),
    },
    // nDCG rather than recall@3 drives the comparisons: it handles several relevant chunks
    // without a ceiling that moves per query, so two queries stay comparable after pooling.
    perQuery: ndcg,
  }
}

export function fmt(x: number, n: number): string {
  // Never dress up a tiny sample as a percentage.
  return n < 10 ? `${x.toFixed(2)}*` : x.toFixed(3)
}

function main(): void {
  const chunks = readJsonLines<Chunk>(CORPUS, { required: true })
  const allLabels = readJsonLines<Labelled>('evals/data/retrieval.labels.jsonl', { required: true })
  const { usable: labels, stale } = partitionLabels(allLabels, chunks)
  if (stale > 0) {
    console.log(`WARNING: ${stale} of ${allLabels.length} labels point at chunks that no longer exist.`)
    console.log('Excluded from scoring — those exchanges changed since labelling. Re-label them.')
  }
  const retrievers = [
    keywordRetriever(chunks),
    bm25Retriever(chunks),
    bm25Retriever(chunks, { name: 'bm25-stem', tokenizer: tokenizeStemmed }),
  ]
  // Optional: present only when the fastText vectors have been extracted, so the eval still runs
  // for anyone without a 1.2 GB download.
  if (existsSync('evals/data/fasttext-vectors.json')) {
    const ft = JSON.parse(readFileSync('evals/data/fasttext-vectors.json', 'utf8')) as {
      dim: number
      vectors: Record<string, number[]>
    }
    retrievers.push(fasttextRetriever(chunks, ft.vectors, ft.dim))
  }
  if (existsSync('evals/data/embeddings.bin') && existsSync('evals/data/query-vectors.json')) {
    const meta = JSON.parse(readFileSync('evals/data/embeddings.index.json', 'utf8')) as EmbeddingsIndex
    const raw = readFileSync('evals/data/embeddings.bin')
    const vectors = unpackVectors(raw, meta.dim, meta.ids.length)
    // Query vectors are precomputed too, so the run stays offline and repeatable.
    const qv = JSON.parse(readFileSync('evals/data/query-vectors.json', 'utf8')) as Record<string, number[]>
    // The freshness checks compare ids, order and a hash of the exact embedded texts, so they are
    // given the same expansion the embedder used rather than being taught what chunking is.
    const cfg = chunkConfigFromEnv()
    const embedded = expandChunks(chunks, cfg)
    assertDenseInputsFresh(
      embedded,
      meta,
      labels.map((l) => l.query),
      qv,
    )
    const dense = denseRetriever(meta.ids, vectors, (q) => new Float32Array(qv[q]))
    // Labels name whole exchanges, so a chunked run is scored on its parents. Named for the split
    // it used: two rows called `bge-m3` in one results file would be unreadable.
    retrievers.push(cfg ? byParent(dense, embedded.length, `bge-m3/${cfg.chars}+${cfg.overlap}`) : dense)
  }
  const kinds: Array<'bulk' | 'hard'> = ['bulk', 'hard']

  const rows: Row[] = []
  const perQuery = new Map<string, number[]>()
  for (const r of retrievers) {
    for (const kind of kinds) {
      const items = labels.filter((l) => l.kind === kind)
      if (items.length === 0) {
        continue
      }
      const { row, perQuery: pq } = score(r, items)
      rows.push({ retriever: r.name, kind, ...row })
      perQuery.set(`${r.name}:${kind}`, pq)
    }
  }

  console.log(`corpus: ${chunks.length} chunks | labels: ${labels.length}`)
  // nDCG and MRR lead because they survive pooling; the recall columns follow with the ceiling
  // that pooling imposed on them.
  console.log('retriever  kind   n   ndcg@10 mrr    r@10   r@3    (max)  r@1    p50ms  p95ms')
  for (const r of rows) {
    console.log(
      `${r.retriever.padEnd(10)} ${r.kind.padEnd(6)} ${String(r.n).padStart(3)} ` +
        [r.ndcg10, r.mrr, r.r10, r.r3].map((v) => fmt(v, r.n).padEnd(6)).join(' ') +
        ` ${r.r3max.toFixed(3).padEnd(6)} ${fmt(r.r1, r.n).padEnd(6)}` +
        ` ${r.p50ms.toFixed(2).padEnd(6)} ${r.p95ms.toFixed(2)}`,
    )
  }
  console.log('(max) = highest r@3 reachable — recall divides by the relevant count, so pooling caps it')
  if (rows.some((r) => r.n < 10)) {
    console.log('* n < 10 — raw indication only, not a measurement')
  }

  reportSplit(
    'edited',
    labels.filter((l) => l.edited),
    labels.filter((l) => !l.edited),
    retrievers,
    'edited scoring higher is evidence the edits leaked, not that they were better',
  )
  const hard = labels.filter((l) => l.kind === 'hard')
  reportSplit(
    'search-located',
    hard.filter((l) => l.via === 'search'),
    hard.filter((l) => l.via === 'browse'),
    retrievers,
    'search-located hard targets contain the words that found them — a gap here is that, not skill',
  )

  // The intervals belong in the committed artifact next to the point estimates: a row read
  // without its CI is what turns a noise-sized gap into a claim.
  const comparisons: Comparison[] = []
  for (const [better, worse] of [
    ['bm25', 'keyword'],
    ['bm25-stem', 'bm25'],
    ['fasttext', 'bm25-stem'],
    ['bge-m3', 'fasttext'],
  ] as const) {
    for (const kind of kinds) {
      const a = perQuery.get(`${better}:${kind}`)
      const b = perQuery.get(`${worse}:${kind}`)
      if (!a || !b) {
        continue
      }
      const { meanDiff, lower, upper } = pairedBootstrap(a, b, 10_000, 1)
      comparisons.push({ better, worse, kind, metric: 'ndcg@10', n: a.length, meanDiff, lower, upper })
      console.log(
        `${better} - ${worse} on ${kind} (ndcg@10): ${meanDiff.toFixed(3)} [${lower.toFixed(3)}, ${upper.toFixed(3)}]`,
      )
    }
  }

  mkdirSync('evals/results', { recursive: true })
  writeFileSync(
    'evals/results/latest-retrieval.json',
    `${JSON.stringify({ corpus: chunks.length, rows, comparisons }, null, 2)}\n`,
  )
}

if (process.argv[1]?.endsWith('run-retrieval.ts')) {
  main()
}

File: evals/run_retrieval.py
import argparse
import json
import sys
import time
from collections.abc import Callable, Mapping, Sequence
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Literal, NamedTuple, NotRequired, TypedDict

from lib.chunk import HasIdText, chunk_config_from_env, expand_chunks
from lib.corpus import CORPUS, Chunk, EmbeddingsIndex, corpus_text_hash
from lib.jsonl import read_json_lines
from lib.metrics import mean, ndcg_at_k, paired_bootstrap, recall_at_k, reciprocal_rank
from lib.paths import DATA, RESULTS
from lib.stem import tokenize_stemmed
from lib.vectors import unpack_vectors
from retrievers.bm25 import BM25Retriever
from retrievers.by_parent import ByParent
from retrievers.dense import DenseRetriever
from retrievers.fasttext import FasttextRetriever
from retrievers.keyword import KeywordRetriever, Retriever

Kind = Literal['bulk', 'hard']
KINDS: tuple[Kind, ...] = ('bulk', 'hard')


class Labelled(TypedDict):
    query: str
    relevant: list[str]
    kind: Kind
    edited: NotRequired[bool]
    via: NotRequired[Literal['search', 'browse']]


class Partitioned(NamedTuple):
    usable: list[Labelled]
    stale: int


def partition_labels(labels: list[Labelled], chunks: Sequence[HasIdText]) -> Partitioned:
    ids = {c.id for c in chunks}
    usable = [label for label in labels if all(id in ids for id in label['relevant'])]
    return Partitioned(usable, len(labels) - len(usable))


@dataclass(frozen=True, slots=True)
class Stats:
    n: int
    r1: float
    r3: float
    r3max: float
    r10: float
    mrr: float
    ndcg10: float
    p50ms: float
    p95ms: float


@dataclass(frozen=True, slots=True)
class Row(Stats):
    retriever: str
    kind: str


class Scored(NamedTuple):
    row: Stats
    per_query: list[float]


REMEDY = 're-run pnpm eval:embed'


def assert_dense_inputs_fresh(
    chunks: Sequence[HasIdText], meta: EmbeddingsIndex, queries: list[str], qv: Mapping[str, Sequence[float]]
) -> None:
    corpus_ids = [c.id for c in chunks]
    corpus = set(corpus_ids)
    embedded = set(meta['ids'])
    missing = [id for id in corpus_ids if id not in embedded]
    stale = [id for id in meta['ids'] if id not in corpus]
    if missing or stale:
        raise ValueError(
            f'embeddings.index.json does not match the corpus: {len(missing)} chunk(s) unembedded, '
            f'{len(stale)} embedded id(s) no longer in the corpus — {REMEDY}'
        )
    text_hash = meta.get('textHash')
    if text_hash != corpus_text_hash(chunks):
        why = 'hash mismatch' if text_hash else 'no hash recorded'
        raise ValueError(
            f'embeddings.index.json was built from different chunk text (same ids, {why}) — {REMEDY}'
        )
    bad = [q for q in queries if len(qv.get(q, ())) != meta['dim']]
    if bad:
        raise ValueError(
            f'{len(bad)}/{len(queries)} labelled queries have no vector of dim {meta["dim"]} '
            f'(first: {json.dumps(bad[0], ensure_ascii=False)}) — {REMEDY}'
        )


def score(r: Retriever, items: list[Labelled]) -> Scored:
    times: list[float] = []
    ranked: list[list[str]] = []
    for it in items:
        t0 = time.perf_counter()
        ranked.append(r.search(it['query'], 10))
        times.append((time.perf_counter() - t0) * 1000)
    times.sort()

    def pct(p: float) -> float:
        return times[min(len(times) - 1, int(p * len(times)))] if times else 0

    def each(metric: Callable[[list[str], list[str]], float]) -> float:
        return mean([metric(rk, it['relevant']) for rk, it in zip(ranked, items, strict=True)])

    ndcg = [ndcg_at_k(rk, it['relevant'], 10) for rk, it in zip(ranked, items, strict=True)]
    row = Stats(
        n=len(items),
        r1=each(lambda rk, rel: recall_at_k(rk, rel, 1)),
        r3=each(lambda rk, rel: recall_at_k(rk, rel, 3)),
        r3max=mean([min(3, len(it['relevant'])) / len(it['relevant']) for it in items]),
        r10=each(lambda rk, rel: recall_at_k(rk, rel, 10)),
        mrr=each(reciprocal_rank),
        ndcg10=mean(ndcg),
        p50ms=pct(0.5),
        p95ms=pct(0.95),
    )
    return Scored(row, ndcg)


def fmt(x: float, n: int) -> str:
    return f'{x:.2f}*' if n < 10 else f'{x:.3f}'


def report_split(
    name: str, subset: list[Labelled], rest: list[Labelled], retrievers: Sequence[Retriever], warning: str
) -> None:
    if not subset or not rest:
        return
    print(f'\n{name}: {len(subset)} of {len(subset) + len(rest)}')
    for r in retrievers:
        a = score(r, subset).row
        b = score(r, rest).row
        print(
            f'{r.name:<10} ndcg@10 {name} {fmt(a.ndcg10, len(subset))} (n={len(subset)}) '
            f'vs rest {fmt(b.ndcg10, len(rest))} (n={len(rest)})'
        )
    print(warning)


def load_retrievers(chunks: list[Chunk], labels: list[Labelled]) -> list[Retriever]:
    retrievers: list[Retriever] = [
        KeywordRetriever(chunks),
        BM25Retriever(chunks),
        BM25Retriever(chunks, name='bm25-stem', tokenizer=tokenize_stemmed),
    ]
    fasttext_path = DATA / 'fasttext-vectors.json'
    if fasttext_path.exists():
        ft = json.loads(fasttext_path.read_text(encoding='utf-8'))
        retrievers.append(FasttextRetriever(chunks, ft['vectors'], ft['dim']))
    if (DATA / 'embeddings.bin').exists() and (DATA / 'query-vectors.json').exists():
        meta: EmbeddingsIndex = json.loads((DATA / 'embeddings.index.json').read_text(encoding='utf-8'))
        vectors = unpack_vectors((DATA / 'embeddings.bin').read_bytes(), meta['dim'], len(meta['ids']))
        qv: dict[str, list[float]] = json.loads((DATA / 'query-vectors.json').read_text(encoding='utf-8'))
        cfg = chunk_config_from_env()
        embedded = expand_chunks(chunks, cfg)
        assert_dense_inputs_fresh(embedded, meta, [label['query'] for label in labels], qv)
        dense = DenseRetriever(meta['ids'], vectors, lambda q: qv[q])
        retrievers.append(
            ByParent(dense, len(embedded), f'bge-m3/{cfg.chars}+{cfg.overlap}') if cfg else dense
        )
    return retrievers


COMPARISONS = (('bm25', 'keyword'), ('bm25-stem', 'bm25'), ('fasttext', 'bm25-stem'), ('bge-m3', 'fasttext'))


def main(out: Path) -> None:
    chunks = [Chunk(**c) for c in read_json_lines(CORPUS, required=True)]
    all_labels: list[Labelled] = read_json_lines(DATA / 'retrieval.labels.jsonl', required=True)
    labels, stale = partition_labels(all_labels, chunks)
    if stale > 0:
        print(f'WARNING: {stale} of {len(all_labels)} labels point at chunks that no longer exist.')
        print('Excluded from scoring — those exchanges changed since labelling. Re-label them.')
    retrievers = load_retrievers(chunks, labels)

    rows: list[Row] = []
    per_query: dict[str, list[float]] = {}
    for r in retrievers:
        for kind in KINDS:
            items = [label for label in labels if label['kind'] == kind]
            if not items:
                continue
            scored = score(r, items)
            rows.append(Row(retriever=r.name, kind=kind, **asdict(scored.row)))
            per_query[f'{r.name}:{kind}'] = scored.per_query

    print(f'corpus: {len(chunks)} chunks | labels: {len(labels)}')
    print('retriever  kind   n   ndcg@10 mrr    r@10   r@3    (max)  r@1    p50ms  p95ms')
    for row in rows:
        print(
            f'{row.retriever:<10} {row.kind:<6} {row.n:>3} '
            + ' '.join(f'{fmt(v, row.n):<6}' for v in (row.ndcg10, row.mrr, row.r10, row.r3))
            + f' {row.r3max:<6.3f} {fmt(row.r1, row.n):<6}'
            + f' {row.p50ms:<6.2f} {row.p95ms:.2f}'
        )
    print('(max) = highest r@3 reachable — recall divides by the relevant count, so pooling caps it')
    if any(row.n < 10 for row in rows):
        print('* n < 10 — raw indication only, not a measurement')

    report_split(
        'edited',
        [label for label in labels if label.get('edited')],
        [label for label in labels if not label.get('edited')],
        retrievers,
        'edited scoring higher is evidence the edits leaked, not that they were better',
    )
    hard = [label for label in labels if label['kind'] == 'hard']
    report_split(
        'search-located',
        [label for label in hard if label.get('via') == 'search'],
        [label for label in hard if label.get('via') == 'browse'],
        retrievers,
        'search-located hard targets contain the words that found them — a gap here is that, not skill',
    )

    comparisons: list[dict[str, str | int | float]] = []
    for better, worse in COMPARISONS:
        for kind in KINDS:
            a = per_query.get(f'{better}:{kind}')
            b = per_query.get(f'{worse}:{kind}')
            if a is None or b is None:
                continue
            ci = paired_bootstrap(a, b, 10_000, 1)
            comparisons.append(
                {
                    'better': better,
                    'worse': worse,
                    'kind': kind,
                    'metric': 'ndcg@10',
                    'n': len(a),
                    'meanDiff': ci.mean_diff,
                    'lower': ci.lower,
                    'upper': ci.upper,
                }
            )
            print(
                f'{better} - {worse} on {kind} (ndcg@10): {ci.mean_diff:.3f} [{ci.lower:.3f}, {ci.upper:.3f}]'
            )

    out.parent.mkdir(parents=True, exist_ok=True)
    report = {'corpus': len(chunks), 'rows': [asdict(row) for row in rows], 'comparisons': comparisons}
    out.write_text(json.dumps(report, indent=2, ensure_ascii=False) + '\n', encoding='utf-8')


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Score every retriever over the labelled queries')
    parser.add_argument('--out', type=Path, default=RESULTS / 'latest-retrieval.json')
    args = parser.parse_args(sys.argv[1:])
    main(args.out)
